ballerina/utils.bal

```diff
@@ -0,0 +1,115 @@
+        return <persist:Error>error("Unable to determine key from DuplicateKey error message.");
+    }
+
+    string key = errorMessage.substring(startIndex + 18, endIndex);
+    return key;
+}
+
+isolated function convertToArray(typedesc<record {}> elementType, record {}[] arr) returns elementType[] = @java:Method {
+    'class: "io.ballerina.stdlib.persist.googlesheets.Utils"
+} external;
+
+
+# Closes the entity stream.
+#
+# + customStream - Stream that needs to be closed
+# + return - `()` if the operation is performed successfully or a `persist:Error` if the operation fails
+public isolated function closeEntityStream(stream<anydata, error?>? customStream) returns persist:Error? {
+    if customStream is stream<anydata, error?> {
+        error? e = customStream.close();
+        if e is error {
+            return <persist:Error>error(e.message());
+        }
+    }
+}
+
+public isolated function filterRecord(record {} 'object, string[] fields) returns record {} {
+    record {} retrieved = {};
+
+    foreach string 'field in fields {
+
+        // ignore many relations
+        if 'field.includes("[]") {
+            continue;
+        }
+
+        // if field is part of a relation
+        if 'field.includes(".") {
+
+            int splitIndex = <int>'field.indexOf(".");
+            string relation = 'field.substring(0, splitIndex);
+            string innerField = 'field.substring(splitIndex + 1, 'field.length());
+
+            if 'object[relation] is record {} {
+                anydata val = (<record {}>'object[relation])[innerField];
+
+                if !(retrieved[relation] is record {}) {
+                    retrieved[relation] = {};
+                }
+
+                record {} innerRecord = <record {}>'retrieved[relation];
+                innerRecord[innerField] = val;
+            }
+        } else {
+            retrieved['field] = 'object['field];
+        }
+
+    }
+    return retrieved;
+}
+
+public isolated function getSheetIds(sheets:Client googleSheetClient, record {|SheetMetadata...;|} metadata, string spreadsheetId) returns map<int>|persist:Error {
+    map<int> sheetIds = {};
+    sheets:Sheet|error sheet;
+    foreach string key in metadata.keys() {
+        sheet = googleSheetClient->getSheetByName(spreadsheetId, metadata.get(key).tableName);
+        if sheet is error {
+            return <persist:Error>error(sheet.message());
+        }
+        sheetIds[key] = sheet.properties.sheetId;
+    }
+    return sheetIds;
+}
+
+public isolated function getKey(anydata|record {} 'object, string[] keyFields) returns anydata|record {} {
+    record {} keyRecord = {};
+
+    if keyFields.length() == 1 && 'object is record {} {
+        return 'object[keyFields[0]];
+    }
+
+    if 'object is record {} {
+        foreach string key in keyFields {
+            keyRecord[key] = 'object[key];
+        }
+    } else {
+        keyRecord[keyFields[0]] = 'object;
+    }
+    return keyRecord;
+}
```

Comment:
We can use the same util method from the persist module

Reply:
updated in [3a96dac](https://github.com/ballerina-platform/module-ballerinax-persist.googlesheets/pull/1/commits/3a96dacf553e65a45bff651098a8c17195b72230)